### DISCUSSION TITLE
Should indexes start at 0 or 1?

### DISCUSSION THREAD

**Discussion by lmorg** [lmorg/murex#878] [Ideas]
When writing a document on working with arrays, I realized just how inconsistent the rules were about where indexes start from:

https://dev.murex.rocks/user-guide/murex-arrays.html

- indexes start from 0
- ranges start from 1
- negative indexes start from 1

It's all a bit confusing and inconsistent. I'm thinking from version v7.0, I'll bring everything in line.

I can do this by having a feature flag: `runmode version 7.0`, if that's included in a module or script, then the code will conform to the new standard. If it's not included then code will respect the current design albeit with a log line written to something like:

```
~/.murex_modules/deprecation_notices.log
```

...so that people can review and update any old code.

---

As an aside, I'm thinking this `runmode version ...` concept and logging could be a good way to bring in changes gracefully. ie allow legacy code to continue to work but allow improvement to the shell to proceed unencumbered.

I'd be interested in peoples thoughts about this.

**Comment by tiymat** [comment]
I've been busy of late and am only seeing this a month+ later, but I agree that having different structures/types using different indexing is incongruous!

Just to be clear, by `bring everything in line` did you mean making indices 1-indexed or ranges 0-indexed? From the title I guess the former? 

I like the `runmode` idea, it's a nice way to maintain backwards compatibility without impeding the implementation of new features! My only real question is, would it only be usable in this way at the module/script level? I'm imagining a situation, admittedly an edge case, where the module starts with `runmode version X` and it has a function that defines a different runmode version.

**Reply by lmorg** [reply]
>Just to be clear, by bring everything in line did you mean making indices 1-indexed or ranges 0-indexed? From the title I guess the former?
I haven't really decided to be honest. There's strong arguments in favour of either:
0. more common with more traditional programming languages
1. more common with shell languages

When working with (for example) JSON arrays you'd expect the first index to be 0. But when counting columns you'd expect it to be 1 (like with `awk`, `sort`, etc)

I don't really know what's best here.

> I like the runmode idea, it's a nice way to maintain backwards compatibility without impeding the implementation of new features! My only real question is, would it only be usable in this way at the module/script level? I'm imagining a situation, admittedly an edge case, where the module starts with runmode version X and it has a function that defines a different runmode version.

I've had a prototype version of the code running for a few weeks now and it's decided by the `json` manifests in the modules directory. This means it is at a module level and not set by code. Which has several disadvantages:
- doesn't support function scoping
- doesn't support individual scripts
- doesn't support `.murex_profile`
However the code is working and working pretty well. So it's a good foundation to build from.

**Comment by tiymat** [comment]
While poking at something today I ran into the index/range 0/1 difference and it definitely took me by surprise!

**Reply by lmorg** [reply]
yeah, it sucks. Even I hate it. I'm really not sure what the best approach is here.

**Comment by jcrben** [comment]
strong +1 for starting at 0
